### DISCUSSION TITLE
'before_connect' Event

### DISCUSSION THREAD

**Discussion by AutumnSun1996** [sqlalchemy/sqlalchemy#6352] [Usage Questions] [answered]
I'm using sqlalchemy to connect to a bunch of database, including a impala server.
We use kerberos for auth, so I need to call os.system('kinit') before any connection attempt to impala.

Sqlalchemy has event system, however there seems to be no 'before_connect' event.

This problem can be solved with custom `creator` function, but it will result in inconsistent connection params.

With a 'before_connect' event, it should look like:
```python

def kinit(*args, **kw):
    os.system(KINIT_CMD)

if need_kerberos(engine):
    listen(engine, 'before_connect', kinit)
```

And here is the hack I am using now:
```python
def with_kinit(func):
    @wraps(func)
    def wrapped(*args, **kwars):
        os.system(KINIT_CMD)
        return func(*args, **kwars)

    return wrapped

if need_kerberos(engine):
    engine.connect = with_kinit(engine.connect)
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

You may be able to use this event: [DialectEvents.do_connect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=event#sqlalchemy.events.DialectEvents.do_connect)

**Comment by zzzeek** [comment]
hey there -

we have a lot of docs for customization of connectivity here: https://docs.sqlalchemy.org/en/14/core/engines.html#custom-dbapi-connect-arguments-on-connect-routines
